File: java/src/dust/crypto/Curve.java
package dust.crypto;

import djb.Curve25519;
import dust.crypto.Keypair;

public class Curve
{
  static public Keypair createKeypair(byte[] entropy)
  {
    byte[] pub=new byte[32];
    byte[] secret=new byte[32];

    System.arraycopy(entropy, 0, secret, 0, 32);

    Curve25519.keygen(pub, null, secret);

    return new Keypair(new Key(secret), new Key(pub));
  }

  static public byte[] createShared(byte[] secret, byte[] pubkey2)
  {
    byte[] shared=new byte[32];

    Curve25519.curve(shared, secret, pubkey2);

    return shared;
  }
}

File: hs/Dust/Crypto/Curve.hs
module Dust.Crypto.Curve
(
  splitSecret,
  createPrivate,
  createPublic,
  createKeypair,
  createShared,
  createEphemeral
) where

import Data.ByteString as B
import Data.Word
import Data.Bits
import System.Entropy

import Dust.Crypto.Keys
import Dust.Crypto.DustCipher
import Dust.Crypto.Curve25519

createEphemeral :: IO (Keypair)
createEphemeral = do
    entropy <- getEntropy 32
    return (createKeypair entropy)

splitSecret :: ByteString -> (Word8,ByteString,Word8)
splitSecret bs = let firstByte = B.head bs
                     lastByte  = B.last bs
                     middle = B.tail (B.init bs)
                 in (firstByte,middle,lastByte)

createPrivate :: ByteString -> PrivateKey
createPrivate bs = let (firstByte,middle,lastByte) = splitSecret bs
                       firstByte' = firstByte .&. 248
                       lastByte'  = (lastByte .&. 127) .|. 64
                   in PrivateKey (firstByte' `cons` middle `snoc` lastByte')

createPublic :: PrivateKey -> PublicKey
createPublic private = let bps = pack [9,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
                       in PublicKey (curve25519 (privateBytes private) bps)

createKeypair :: ByteString -> Keypair
createKeypair entropy = let private = createPrivate entropy
                            public = createPublic private
                        in Keypair public private

createShared :: PrivateKey -> PublicKey -> EncryptionKey
createShared private public = EncryptionKey (curve25519 (privateBytes private) (publicBytes public))
